pkg/controller/ephemeraljob/ephemeraljob_controller.go

```diff
@@ -0,0 +1,334 @@
+	// Create a new controller
+	c, err := controller.New("ephemeraljob-controller", mgr, controller.Options{Reconciler: r, MaxConcurrentReconciles: concurrentReconciles})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to EphemeralJob
+	err = c.Watch(&source.Kind{Type: &appsv1alpha1.EphemeralJob{}}, &handler.EnqueueRequestForObject{}, predicate.Funcs{
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			old := e.ObjectOld.(*appsv1alpha1.EphemeralJob)
+			new := e.ObjectNew.(*appsv1alpha1.EphemeralJob)
+			// only update these fields
+			if old.Spec.TTLSecondsAfterCreated != new.Spec.TTLSecondsAfterCreated ||
+				old.Spec.Paused != new.Spec.Paused || old.Spec.Parallelism != new.Spec.Parallelism ||
+				old.Spec.Replicas != new.Spec.Replicas {
+
+				klog.V(3).Infof("Observed updated Spec for ephemeral job: %s/%s", new.Namespace, new.Name)
+				return true
+			}
+			return false
+		},
+	})
+	if err != nil {
+		return err
+	}
+	// Watch for changes to Pod
+	err = c.Watch(&source.Kind{Type: &v1.Pod{}}, &podHandler{mgr.GetCache()})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+var _ reconcile.Reconciler = &ReconcileEphemeralJob{}
+
+// ReconcileImagePullJob reconciles a ImagePullJob object
+type ReconcileEphemeralJob struct {
+	client.Client
+	scheme *runtime.Scheme
+}
+
+// +kubebuilder:rbac:groups=apps.kruise.io,resources=ephemeraljobs,verbs=get;list;watch;update;patch;delete
+// +kubebuilder:rbac:groups=apps.kruise.io,resources=ephemeraljobs/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=core,resources=pods,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=core,resources=pods/status,verbs=get
+// +kubebuilder:rbac:groups=core,resources=pods/ephemeralcontainers,verbs=get;update
+
+// Reconcile reads that state of the cluster for a EphemeralJob object and makes changes based on the state read
+// and what is in the EphemeralJob.Spec
+func (r *ReconcileEphemeralJob) Reconcile(context context.Context, request reconcile.Request) (res reconcile.Result, err error) {
+	start := time.Now()
+	klog.V(5).Infof("Starting to process EphemeralJob %v", request.Name)
+	var requeueMsg string
+	defer func() {
+		if err != nil {
+			klog.Warningf("Failed to process EphemeralJob %v err %v, elapsedTime %v", request.Name, time.Since(start), err)
+		} else if res.RequeueAfter > 0 {
+			klog.Infof("Finish to process EphemeralJob %v, elapsedTime %v, RetryAfter %v: %v", request.Name, time.Since(start), res.RequeueAfter, requeueMsg)
+		} else {
+			klog.Infof("Finish to process EphemeralJob %v, elapsedTime %v", request.Name, time.Since(start))
+		}
+	}()
+
+	job := &appsv1alpha1.EphemeralJob{}
+	err = r.Get(context, request.NamespacedName, job)
+
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Object not found, return.  Created objects are automatically garbage collected.
+			// For additional cleanup logic use finalizers.
+			return reconcile.Result{}, nil
```

Comment:
we should delete the expectation for key of job, otherwise if we create a new job with the same name, the expectation is dirty.